SDK Extension Support

tplooker opened the pull request:
This PR adds the ability to easily extend the SDK by overriding the default implementation of the core SDK services while preserving the base line interface and DI resolution for the core services.

- Renames all base service interfaces to be suffix with IDefault.
- Adds a ServicesBuilder class that handles building the SDK services pool.
- Added the ability to control the services injection configuration from the AddAgent extension method.

tmarkovski commented:
> * R.e your first point, my motivation for renaming the interfaces to be generic is that the interface itself could be extended for example we could have `ICustomConnectionService : IDefaultConnectionService`. The approach to the service registration means a custom connection service implementation i.e `CustomConnectionService : ICustomConnectionService` could be resolved to both `IDefaultConnectionService` and `ICustomConnectionService` meaning the existing untouched SDK components relying on the base connection service can resolve the implementation by the base interface, while higher level application components can resolve the extended implementation via its extended interface. The other approach here would be to just add the implementation of the extended service so it can be resolved directly without using an interface.

I see your point. Here's a thought: when you want to create interface hierarchy, it's a lot clearer if you take an approach where you append a prefix to each added super interface. Ex. IConnectionService, ICustomConnectionService, IExtendedCustomConnectionService, etc.
If my class DefaultConnectionService, implements IDefaultConnectionService, then my CustomConnectionService requires me to create an interface ICustomConnectionService just for the sake of consistency, where by DefaultConnectionService and CustomConnectionService both implement IConnectionService, it's very immediately obvious that both are implementations, with Default suggesting something provided by the SDK.
I noticed you also added Default suffix to some of the model classes. Seems like a lot of default notations, it feels little redundant. Classes in C# are inheritable by nature, no need to designate that in the name (exceptions being abstract classes and similar).

> * R.e your second point, just checking I understand what you mean, basically the custom service resolution would occur in the startup.cs explicitly like `services.AddSingleton<IDefaultConnectionService,CustomConnectionService>`, then the service builder essentially just fills in the blanks with try add? I'll take a closer look at the project, thanks!

Correct. These would be the 2 use cases

Case 1:
User registers default implementation using `service.AddAgent()`
 - All default services are added with `TryAddSingleton` (or Transient)

Case 2:
User wants to override ConnectionService using 
```
service.AddAgent(builder =>
{
    build.AddConnectionService<CustomConnectionService>();
});
```
- The extension method `AddConnectionService()` uses `service.AddSingleton`.
- After builder method is executed, we continue adding all services as in Case 1, only using `TryAdd` to make sure the overrides aren't overriden.

From the SDK: Using `AddSingleton` twice will make the dotnet runtime use the latest registration, overriding anything before. This means even if we run our implementation before the builder extension method, service order would be fine, since the extension method `AddConnectionService` won't use `TryAdd`.

tplooker commented:
> > * R.e your first point, my motivation for renaming the interfaces to be generic is that the interface itself could be extended for example we could have `ICustomConnectionService : IDefaultConnectionService`. The approach to the service registration means a custom connection service implementation i.e `CustomConnectionService : ICustomConnectionService` could be resolved to both `IDefaultConnectionService` and `ICustomConnectionService` meaning the existing untouched SDK components relying on the base connection service can resolve the implementation by the base interface, while higher level application components can resolve the extended implementation via its extended interface. The other approach here would be to just add the implementation of the extended service so it can be resolved directly without using an interface.
> 
> I see your point. Here's a thought: when you want to create interface hierarchy, it's a lot clearer if you take an approach where you append a prefix to each added super interface. Ex. IConnectionService, ICustomConnectionService, IExtendedCustomConnectionService, etc.
> In the case you're describing, if my class DefaultConnectionService, implements IDefaultConnectionService, then my CustomConnectionService requires me to create an interface ICustomConnectionService just for the sake of consistency, where by DefaultConnectionService and CustomConnectionService both implement IConnectionService, it's very immediately obvious that both are implementations, with Default suggesting something provided by the SDK.
> I noticed you also added Default suffix to some of the model classes. Seems like a lot of default notations, it feels little redundant. Classes in C# are inheritable by nature, no need to designate that in the name (exceptions being abstract classes and similar).
> 
> > * R.e your second point, just checking I understand what you mean, basically the custom service resolution would occur in the startup.cs explicitly like `services.AddSingleton<IDefaultConnectionService,CustomConnectionService>`, then the service builder essentially just fills in the blanks with try add? I'll take a closer look at the project, thanks!
> 
> Correct. These would be the 2 use cases
> 
> Case 1:
> User registers default implementation using `service.AddAgent()`
> 
> * All default services are added with `TryAddSingleton` (or Transient)
> 
> Case 2:
> User wants to override ConnectionService using
> 
> ```
> service.AddAgent(builder =>
> {
>     build.AddConnectionService<CustomConnectionService>();
> });
> ```
> * The extension method `AddConnectionService()` uses `service.AddSingleton`.
> * After builder method is executed, we continue adding all services as in Case 1, only using `TryAdd` to make sure the overrides aren't overriden.
> 
> From the SDK: Using `AddSingleton` twice will make the dotnet runtime use the latest registration, overriding anything before. This means even if we run our implementation before the builder extension method, service order would be fine, since the extension method `AddConnectionService` won't use `TryAdd`.

Agree with your point around the interface naming, I have updated the implementation to be closer to that of the IdentityServer project you referenced.
